File: internal/database/repository/api_key.go
//go:build js && wasm
package database_repository

import (
	"database/sql"
	"time"

	database_model "github.com/kodesettings/dbmodule/v2/internal/database/model"
)

type ApiKeyRepo struct {
	db *sql.DB
}

func (h *ApiKeyRepo) FindAll() []database_model.ApiKey {
	rows, err := h.db.Query("SELECT * FROM api_key ORDER BY created_at DESC;")
	if err != nil {
		return []database_model.ApiKey{}
	}

	api_keys := []database_model.ApiKey{}
	for rows.Next() {
		var a database_model.ApiKey
		err = rows.Scan(&a.Id, &a.Key, &a.Permissions, &a.CreatedAt, &a.UpdatedAt)
		if err != nil {
			return []database_model.ApiKey{}
		}
		api_keys = append(api_keys, a)
	}

	return api_keys
}

func (h *ApiKeyRepo) Create(api_key database_model.ApiKey) bool {
	database_model.CreateSuperAdminApiKey();

	now := time.Now().Unix()
	api_key.CreatedAt = uint64(now);
	api_key.UpdatedAt = uint64(now);

	_, err := h.db.Exec("INSERT INTO api_key (createdAt, updatedAt) VALUES (?, ?, ?);",
						api_key.CreatedAt, api_key.UpdatedAt)
	if err != nil {
		return false;
	}

	return true;
}


func (h *ApiKeyRepo) Update(api_key database_model.ApiKey) bool {
	database_model.CreateSuperAdminApiKey();

	now := time.Now().Unix()
	api_key.UpdatedAt = uint64(now);

	_, err := h.db.Exec("UPDATE api_key SET (key = ?, permission = ?, updatedAt = ?) WHERE Id = ?;",
						api_key.Key, api_key.Permissions, api_key.UpdatedAt, api_key.Id)
	if err != nil {
		return false;
	}

	return true;
}

func (h *ApiKeyRepo) Remove(api_key database_model.ApiKey) bool {
	database_model.CreateSuperAdminApiKey();

	_, err := h.db.Exec("DELETE FROM api_key WHERE Id = ?;", api_key.Id)
	if err != nil {
		return false;
	}

	return true;
}

func (h *ApiKeyRepo) FindById(id uint64) (database_model.ApiKey, bool) {
	database_model.CreateSuperAdminApiKey();

	row := h.db.QueryRow("SELECT * FROM api_key WHERE Id = ?;", id)

	api_key := database_model.ApiKey{}
	if err := row.Scan(&api_key.Id, &api_key.Key, &api_key.Permissions,
					&api_key.CreatedAt, &api_key.UpdatedAt); err != nil {
		return database_model.ApiKey{}, false;
	}

	return api_key, true;
}

func (h *ApiKeyRepo) FindByKey(key string) (database_model.ApiKey, bool) {
	database_model.CreateSuperAdminApiKey();

	row := h.db.QueryRow("SELECT * FROM api_key WHERE Key = ?;", key)

	api_key := database_model.ApiKey{}
	if err := row.Scan(&api_key.Id, &api_key.Key, &api_key.Permissions,
					&api_key.CreatedAt, &api_key.UpdatedAt); err != nil {
		return database_model.ApiKey{}, false;
	}

	return api_key, true;
}

func (h *ApiKeyRepo) FindByKeyButId(id uint64, key string) (database_model.ApiKey, bool) {
	database_model.CreateSuperAdminApiKey();

	row := h.db.QueryRow("SELECT * FROM api_key WHERE (Id = ?, Key = ?);", id, key)

	api_key := database_model.ApiKey{}
	if err := row.Scan(&api_key.Id, &api_key.Key, &api_key.Permissions,
				&api_key.CreatedAt, &api_key.UpdatedAt); err != nil {
		return database_model.ApiKey{}, false;
	}

	return api_key, true;
}

File: sources/database/repository/api_key.cpp
#include "../model/api_key.hpp"

ApiKey* findAllApiKeys() {
  // TODO: return all rows from the table
  // eg.: SELECT * FROM api_key ORDER BY created_at DESC;
  return new ApiKey[10];
}

bool createApiKey(ApiKey apiKey) {
  createSuperAdminApiKey();

  auto timestamp = std::chrono::system_clock::now().time_since_epoch();
  auto localtime = std::chrono::duration_cast<std::chrono::seconds>(timestamp).count();
  apiKey.createdAt = localtime;
  apiKey.updatedAt = localtime;

  // TODO: create an entry in the table and return it's object
  // eg.: INSERT INTO api_key (createdAt, updatedAt) VALUES (?, ?, ?);

  return true;
}

bool updateApiKey(ApiKey apiKey) {
  createSuperAdminApiKey();

  auto timestamp = std::chrono::system_clock::now().time_since_epoch();
  auto localtime = std::chrono::duration_cast<std::chrono::seconds>(timestamp).count();
  apiKey.updatedAt = localtime;

  // TODO: update entry in table
  // eg.: UPDATE api_key SET (key = ?, permission = ?, updatedAt = ?) WHERE Id = ?;

  return true;
}

bool removeApiKey(uint64 id) {
  createSuperAdminApiKey();

  // TODO: remove entry from table
  // eg.: DELETE FROM api_key WHERE Id = ?;

  return true;
}

ApiKey findApiKeyById(uint64 id, bool *found) {
  createSuperAdminApiKey();

  // TODO: select row by id
  // eg.: SELECT * FROM api_key WHERE Id = ?;
  return ApiKey{};
}

ApiKey findApiKeyByKey(std::string key, bool *found) {
  createSuperAdminApiKey();

  // TODO: select row by key
  // eg.: SELECT * FROM api_key WHERE Key = ?;
  return ApiKey{};
}

ApiKey findApiKeyByKeyButId(uint64 id, std::string key, bool *found) {
  createSuperAdminApiKey();

  // TODO: select row by id and key
  // eg.: SELECT * FROM api_key WHERE (Id = ?, Key = ?);
  return ApiKey{};
}
